src/vibrato.rs

```diff
@@ -0,0 +1,457 @@
+use crate::lfo::LFO;
+use crate::ring_buffer::RingBuffer;
+
+#[derive(Debug, Clone)]
+pub enum Error {
+    InvalidValue { name: String, value: f32 },
+}
+/*
+   Vibrato Effect
+   Implementation borrowed from comb_filter.rs from [https://github.com/manodrum/ase-2024/blob/assignment-1/src/comb_filter.rs]
+   The struct (cnstructor) takes in the following parameters:
+   - sample_rate_hz: f32
+   - delay_secs: f32
+   - width_secs: f32
+   - mod_freq_hz: f32
+   - num_channels: usize
+   The struct has the following methods:
+   - new: creates a new instance of the struct
+   - reset: resets the delay buffer and phase
+   - set_param: sets the parameters of the filter
+   - get_param: gets the parameters of the filter
+   - process: processes the input and writes the output to the output buffer
+*/
+pub struct VFilter {
+    sample_rate_hz: f32,
+    delay_secs: f32,
+    width_secs: f32,
+    mod_freq_hz: f32,
+    delay_buffer: Vec<RingBuffer<f32>>,
+    lfo: LFO, // Add LFO here
+    num_channels: usize,
+}
+
+impl VFilter {
+    /// Creates a new instance of the struct VFilter
+    /// Example usage
+    /// ```
+    /// let vibrato = VFilter::new(44100.0, 0.01, 0.005, 5.0, 2);
+    /// assert_eq!(vibrato.sample_rate_hz, 44100.0);
+    /// assert_eq!(vibrato.delay_secs, 0.01);
+    /// assert_eq!(vibrato.width_secs, 0.005);
+    /// assert_eq!(vibrato.mod_freq_hz, 5.0);
+    /// assert_eq!(vibrato.num_channels, 2);
+    /// ```
+    pub fn new(
+        sample_rate_hz: f32,
+        delay_secs: f32,
+        width_secs: f32,
+        mod_freq_hz: f32,
+        num_channels: usize,
+    ) -> Result<Self, Error> {
+        let delay_samples = (delay_secs * sample_rate_hz) as usize;
+        let width_samples = (width_secs * sample_rate_hz) as usize;
+        let capacity = 1 + delay_samples + width_samples * 2;
+        if width_samples > delay_samples {
+            return Err(Error::InvalidValue {
+                name: "width in seconds".to_string(),
+                value: width_secs,
+            });
+        }
+        if (mod_freq_hz - 0.0).abs() < 1e-6 || mod_freq_hz < 0.0 {
+            return Err(Error::InvalidValue {
+                name: "modulation frequency in Hz".to_string(),
+                value: mod_freq_hz,
+            });
+        }
+        let delay_buffer = vec![RingBuffer::<f32>::new(capacity); num_channels];
+        let mut lfo = LFO::new(sample_rate_hz, 1024);
+        lfo.set_frequency(mod_freq_hz);
```

Comment:
I'm not sure if this part actually sets the frequency, see my comment for lfo.rs